### DISCUSSION TITLE
How does bucket in session get refilled?

### DISCUSSION THREAD

**Discussion by nikhil2209** [vladimir-bukhtoyarov/bucket4j#164] [General]
I was looking through the basic usage examples, and the last one for rate limiting is a bit confusing. The filter stores the bucket in the session, and retrieves it whenever a new request comes in. If the bucket is in session, when is the session variable getting updated to refill the bucket?

To my best understanding, the refill must be happening when we try to consume tokens from the bucket, based on last refill time and the time that has passed since then. So tryConsume(x) first refills the bucket, and then assigns tokens to the request (is this correct?). But nowhere do we explicitly set the updated bucket in the session. This is very confusing.

Exactly how does the bucket in the user session get updated?

**Comment by vladimir-bukhtoyarov** [comment]
@nikhil2209 hello,

Bucket is refilled only at time of interaction, there is no a background refill if nobody interact with bucket. 

According to example, it supposed to be used for scenario when you have only one server in production(looks it need to mention this explicitly in the docs). In such deployment scenario updating internal state of bucket does not require to write something to the session again, because it is not required to replicate changes in the bucket state across multiple servers.

If you want to use Bucket4j in topology when you have multiple servers that process user requests then you need to switch to one of [clustered extension of Bucket4j](https://github.com/vladimir-bukhtoyarov/bucket4j#supported-back-ends).

I will rewrite this example soon to avoid confusing in the scope of #69

**Comment by nikhil2209** [comment]
Thank you for responding to my query. But this confuses me even more. Where is the bucket state maintained if not in session? And if bucket state is not maintained in session, why store the bucket in session?

I should clarify the confusion I'm having. I can't figure out how the example works. I have integrated it into my own project, and it is working correctly, but I have no idea why or how. I am looking for a simple explanation of the flow control that's coded in for this library.

So for example, going back to the rate limiting scenario, a simple explanation can be like:-

1. Whenever a request comes in, we check if there is a bucket in this session.
2. If there isn't, we create a new bucket with required parameters and set it in session.
3. If there was a bucket in the session already, we retrieve it and call tryConsume on it to check if the request can be allowed. (If you could expand this point a little more, it would be great.) 
    a) How does tryConsume() know how many tokens to refill before checking for rate limiting? Is there a bucket creation time stored in the session variable, and that creation time is compared to current time to calculate the refill tokens?  
    b) If we are not writing data back to the session variable, how do you know how many tokens the bucket has left on the next tryConsume() call? More specifically, how is the bucket state maintained across successive requests in the same session?

Again, thank you for helping out, this is a very useful library! It would be great if you could clear the doubts above.

**Comment by vladimir-bukhtoyarov** [comment]
All required information is already stored inside the bucket. It remembers:
* Own configuration.
* Last refill time.
* Available tokens.

So when you retrieve the bucket from the servlet session you get fully operable data. 

Maybe it would be better to store the bucket in the Guava/Caffeine cache instead of ServletSession to get better control over execution, as well as better protection from unexpected behavior of servlet session manager.

**Reply by nikhil2209** [reply]
Thank you. I understand now. I have the following code,

```
import io.github.bucket4j.Bandwidth;
import io.github.bucket4j.Bucket;
import javax.servlet.Filter;
import javax.servlet.FilterChain;
import javax.servlet.FilterConfig;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletResponse;

import io.github.bucket4j.Bucket4j;
import io.github.bucket4j.Refill;
import java.time.Duration;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;

public class RateLimitingFilter implements Filter {

    private Bucket createNewBucket() {
        long overdraft = 3;
        Refill refill = Refill.greedy(1, Duration.ofSeconds(1));
        Bandwidth limit = Bandwidth.classic(overdraft, refill);
        return Bucket4j.builder().addLimit(limit).build();
    }

    @Override
    public void init(FilterConfig config) throws ServletException {
        // Get init parameter 
        String testParam = config.getInitParameter("test-param");

        //Print the init parameter 
        //System.out.println("Test Param in NoCacheFilter: " + testParam);
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
            throws java.io.IOException, ServletException {
        HttpServletRequest httpRequest = (HttpServletRequest) request;
        HttpSession session = httpRequest.getSession(true);

        
        Bucket bucket = (Bucket) session.getAttribute("throttler");
        if (bucket == null) {
            bucket = createNewBucket();
            session.setAttribute("throttler", bucket);
        }

        // tryConsume returns false immediately if no tokens available with the bucket
        if (bucket.tryConsume(1)) {
            // the limit is not exceeded
            chain.doFilter(request, response);
        } else {
            // limit is exceeded
            System.out.println("Rate Limiter throttling request from : " + request.getRemoteAddr());
            HttpServletResponse httpResponse = (HttpServletResponse) response;
            httpResponse.setContentType("text/plain");
            httpResponse.setStatus(429);
            httpResponse.getWriter().append("Too many requests");
        }
    }

    @Override
    public void destroy() {
        /* Called before the Filter instance is removed 
      from service by the web container*/
    }
}
```

What I want to accomplish is 1 request per second per user session, to protect against DoS attacks against my application. This will accomplish the rate limiting correctly, right?

**Reply by vladimir-bukhtoyarov** [reply]
I think that it should work. It will work until you have only one server, or it is not critical that client can overconsume the limit by requesting multiple servers in parallel.
